primitives/tf_encrypted/primitives/sodium/cc/ops.cc

```diff
@@ -33,4 +67,15 @@ REGISTER_OP("SodiumEasyBoxOpenDetached")
     .Input("pk_sender: uint8")
     .Input("sk_receiver: uint8")
     .Output("plaintext: plaintext_dtype")
-    .SetIsStateful();
+    .SetIsStateful()
+    .SetShapeFn([](shape_inference::InferenceContext* c){
+        shape_inference::ShapeHandle ciphertext_shape = c->input(0);
+        shape_inference::ShapeHandle plaintext_shape;
+        auto rank = c->Rank(ciphertext_shape);
+        if (rank == 0)
+            plaintext_shape = c->MakeShape({});
```

Comment:
perhaps it's better not to set a shape in this case, ie you could rewrite this as follows:

```
if (rank >= 1) {
  TF_RETURN_IF_ERROR(c->Subshape(ciphertext_shape, 0, rank-1, &plaintext_shape));
  c->set_output(0, plaintext_shape);
}
```

Reply:
no forget what I said; your solution is better 👍

Reply:
Okay maybe the story doesn't end here after all.

AFAICT tensors of shape `()` are scalars, and ciphertext scalars only make sense for scalar `uint8` plaintexts. [This line](https://github.com/tf-encrypted/federated/blob/master/tensorflow_federated/python/core/api/computation_types.py#L101) in TFF seems a bit fishy to me since `tf.TensorShape([]) != tf.TensorShape(None)`, but I could be missing an equivalence here (cc @jvmncs any thoughts?).

At any rate, perhaps it's worth being a bit more explicit about the type signatures for our `tf_computations`, using e.g. `tf.TensorSpec(tf.uint8, None)` instead of simply `tf.uint8`. Let me test if this solves the problem you ran into.

Moreover, perhaps it's better to use `c->RankKnown` instead of testing against `c->Rank()`. In any case, if the rank is unknown then `c->Rank()` returns -1, which will caused problems if we're checking `== 0` instead of flipping branches and using `>= 1`.

Reply:
> but I could be missing an equivalence here

They're definitely not equivalent in general. tf.TensorShape([]) denotes scalar, whereas tf.TensorShape(None) is essentially a shapeless placeholder (e.g. user-defined). It seems to me like this is a restriction TFF introduces that deviates from the TF definition. Note they still allow for passing a `TensorShape(None)` object to that constructor.

Reply:
I think I'm close to being at the bottom of this. Using e.g. `tf.uint8` as a type argument is indicating a scalar, and it should indeed be something like `tf.TensorSpec(dtype=tf.uint8, shape=None)` instead.

Reply:
Yep, in that case I think you'd want something like `tff.TensorType(tf.uint8, tf.TensorShape(None))`

Reply:
yes, or simply `tff.TensorType(tf.uint8, None)` (note that `tff.to_type(tf.TensorSpec(dtype=dtype, shape=shape)) == tff.TensorType(dtype, shape)`